### ISSUE TITLE
Access-Control-Allow-Credentials header is missing in server response with basic authentication enabled

### ISSUE THREAD

**Issue by leoek** [jsreport/jsreport-authentication#4]
I noticed that the server is responding without the `Access-Control-Allow-Credentials:true` header being set, which results in a CORS error if the withCredentials flag is set.
I guess this is intended, because setting the `Access-Control-Allow-Credentials:true` header would result in a CORS error, due to the wildcard in the `Access-Control-Allow-Origin` header.
However this could be fixed if there would be an opportunity to specify these headers in the config.

**Comment by bjrmatos** [comment]
ok, i don't see anything related to CORS in the package so probably those options are missing because it was out of scope at that time.

the options makes sense to me, would you like to send a PR to add those options? i think three new options will be needed for [Access-Control-Allow-Credentials](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials), [Access-Control-Allow-Origin](https://developer.mozilla.org/es/docs/Web/HTTP/Headers/Access-Control-Allow-Origin), and [according to MDN (see section CORS and caching)](https://developer.mozilla.org/es/docs/Web/HTTP/Headers/Access-Control-Allow-Origin) the [Vary](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary) header. probably the options can live under a new config property called `cors`

**Comment by pofider** [comment]
The cors response headers are configured using the cors middleware in the express extension [here](https://github.com/jsreport/jsreport-express/blob/master/lib/reporter.express.js#L125).

We should likely just use option `credentials:true` there.

**Comment by bjrmatos** [comment]
ahh i see, but those kind of options should not be configurable? i mean even disabling cors, or allowing specific origins, etc. so those options makes sense in the jsreport-express package?

**Comment by pofider** [comment]
yes, it can be configurable... the authentication extension could for example automatically add `credentials:true`

**Comment by leoek** [comment]
Ok, I would suggest changing it to the following.
The constants should be retrieved from the config with the defaults being set to the values below. except origin which shouldn't have any default value because it's different for every use case.

```
const origin = (config.authentication.origin) ? config.authentication.origin : true; //["origin1","origin2"]
const allowedHeaders = (config.authentication.allowedHeaders) ? config.authentication.allowedHeaders : ["Content-Type", "Authorization"];
const credentials = config.authentication.enabled;

app.options('*', function (req, res) {
    require('cors')({
      methods: ['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'MERGE'],
      credentials
      origin,
      allowedHeaders
    })(req, res)
  })
```
[Link to the line](https://github.com/jsreport/jsreport-express/blob/master/lib/reporter.express.js#L112)

**Comment by leoek** [comment]
If we are already adding settings, maybe its a could idea to add a realm setting too.

**Comment by bjrmatos** [comment]
> Ok, I would suggest changing it to the following.
> The constants should be retrieved from the config with the defaults being set to the values below. except origin which shouldn't have any default value because it's different for every use case.
> 
> const origin = (config.authentication.origin) ? config.authentication.origin : true; //["origin1","origin2"]
> const allowedHeaders = (config.authentication.allowedHeaders) ? config.authentication.allowedHeaders : ["Content-Type", "Authorization"];
> const credentials = config.authentication.enabled;

@leoek i think CORS options makes more sense to live in the [jsreport-express](https://github.com/jsreport/jsreport-express) than in jsreport-authentication options, since CORS is not just about authentication, it is something more general and at the request level.

for me it also makes sense to match [cors](https://github.com/expressjs/cors) options with our options, as demonstrated in your example, one minor thing is.. i'm not completely sure why we have technically [two](https://github.com/jsreport/jsreport-express/blob/master/lib/reporter.express.js#L113) [different](https://github.com/jsreport/jsreport-express/blob/master/lib/reporter.express.js#L125) options for the two places where CORS is initiated, i would prefer that the two share the configuration options, the only difference seems like the first need the [current methods options](https://github.com/jsreport/jsreport-express/blob/master/lib/reporter.express.js#L114) but after that i think both options should match

> If we are already adding settings, maybe its a could idea to add a realm setting too.

sounds good, this option should live in jsreport-authentication to make the realm response configurable.

**Comment by leoek** [comment]
> @leoek i think CORS options makes more sense to live in the jsreport-express than in jsreport-authentication options, since CORS is not just about authentication, it is something more general and at the request level.

I agree with that, however I am not too familiar with jsreports and its splitting into extensions.
>for me it also makes sense to match cors options with our options, as demonstrated in your example

It makes certainly sense, but if cors is configured outside the authentication config it should be possible to override these values.
> i'm not completely sure why we have technically two different options for the two places where CORS is initiated, i would prefer that the two share the configuration options, the only difference seems like the first need the current methods options but after that i think both options should match

If i understand the second the documentation of the cors package correctly the `app.use(cors())` shouldn't be there as it just enables every cors request, which is not what we want.

**Comment by bjrmatos** [comment]
> I agree with that, however I am not too familiar with jsreports and its splitting into extensions.

sure, no problem. it seems to me that new options about cors should live in [jsreport-express](https://github.com/jsreport/jsreport-express), since this is the extension that is handling the web server and CORS.



> It makes certainly sense, but if cors is configured outside the authentication config it should be possible to override these values.

yes, with the new options CORS handling will be configurable. those new options probably will look like: 

```js
// dev.config.json
{
  ...
  express: {
    ...
    cors: {
      // new options here
    }
    ...
  }
  ...
}
```




> If i understand the second the documentation of the cors package correctly the app.use(cors()) shouldn't be there as it just enables every cors request, which is not what we want.

according to the [cors docs](https://github.com/expressjs/cors#enabling-cors-pre-flight) it seems that both are required, the first for Pre-Flight (handled by express by using app.options) requests, and the second for the normal requests, my point is that both `cors()` calls should share the same configuration (the only difference will be that the first need [methods options](https://github.com/jsreport/jsreport-express/blob/master/lib/reporter.express.js#L114)).

if you feel adventurous you can work in a PR that adds these options in the jsreport-express package, otherwise we will look into this soon. thanks for reporting the enhancement

**Comment by pofider** [comment]
Great! I think you have reached the final spec here. I agree to do it the proposed way as well.
